blog/02-02-2023-on-chain-randomness-with-celo-using-subgraphs/on-chain-randomness-with-celo-using-subgraphs.md

```diff
@@ -0,0 +1,1859 @@
+        );
+        membersLimit = membersLimit_;
+        _updateReward();
+    }
+
+    function claimFee() external onlyManager {
+        require(!lotteryStatus, "LotteryClubToken: Lottery is running");
+        require(
+            _managerFee > 0,
+            "LotteryClubToken: Manager fee is 0"
+        );
+        uint256 _feeAmount = _managerFee;
+        _managerFee = 0;
+        IERC20(rewardAddress).transfer(manager, _feeAmount);
+    }
+
+    function start() external onlyManager {
+        require(!lotteryStatus, "LotteryClubToken: Lottery is running");
+        lotteryStatus = true;
+        _reset();
+    }
+
+    function draw() external onlyManager {
+        require(lotteryStatus, "LotteryClubToken: Lottery is not running");
+        require(
+            _membersCounters.length >= membersLimit,
+            "LotteryClubToken: Not enough members"
+        );
+        require(IERC20(rewardAddress).balanceOf(address(this)) >= reward, "LotteryClubToken: Not enough reward");
+        lotteryStatus = false;
+        winer = _membersCounters[_getRandomNumber() % _membersCounters.length];
+        IERC20(rewardAddress).transfer(winer, reward);
+        emit Winer(winer, address(this));
+    }
+
+    function register() external {
+        require(lotteryStatus, "LotteryClubToken: Lottery is not running");
+        require(
+            _membersCounters.length < membersLimit,
+            "LotteryClubToken: Members limit reached"
+        );
+        require(
+            _userDeposit[msg.sender] == 0,
+            "LotteryClubToken: You are already registered"
+        );
+        IERC20(rewardAddress).transferFrom(msg.sender, address(this), deposit);
+        _userDeposit[msg.sender] = deposit;
+        _membersCounters.push(msg.sender);
+    }
+
+    function _updateReward() private {
+        uint256 baseReward = deposit.mul(membersLimit);
+        _managerFee = baseReward.mul(FEE_PERCENT).div(100);
+        reward = baseReward.sub(_managerFee);
+        emit RewardUpdate(reward, deposit, membersLimit, address(this));
+    }
+
+    function _reset() private {
+        winer = address(0);
+        for (uint256 i = 0; i < _membersCounters.length; i++) {
+            _userDeposit[_membersCounters[i]] = 0;
+        }
+        _membersCounters = new address[](0);
+    }
+
+    function _getRandomNumber() private view returns (uint256) {
+        return uint256(RANDOMNESS_ADDRESS.random());
+    }
+}
+```
+
+The code above represents the main code of the `LotteryClubToken` contract. We will provide a detailed explanation of its contents.
+
+First, we need to define the license and solidity version we will use and import some libraries that we will require. The libraries we will use include `SafeMath` to perform mathematical operations and prevent overflow and underflow, `IRandom` to retrieve random values from the `Celo Randomness contract`, and `IERC20` to interact with the tokens we will use as rewards.
```

Comment:
Where is the Celo Randomness contract, It's not here, and it's causing compiling error

Reply:
To find the Celo Randomness Contract address, you can refer to the state variable that I defined.

```solidity
IRandom private constant RANDOMNESS_ADDRESS =
    IRandom(0xdd318EEF001BB0867Cd5c134496D6cF5Aa32311F);
```